flink-connector-aws-dynamodb/src/main/java/org/apache/flink/streaming/connectors/dynamodb/sink/DynamoDbSinkBuilder.java

```diff
@@ -0,0 +1,134 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.connectors.dynamodb.sink;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.connector.base.sink.AsyncSinkBaseBuilder;
+import org.apache.flink.connector.base.sink.writer.ElementConverter;
+import org.apache.flink.streaming.connectors.dynamodb.config.DynamoDbTablesConfig;
+
+import java.util.Optional;
+import java.util.Properties;
+
+/**
+ * Builder to construct {@link DynamoDbSink}.
+ *
+ * <p>The following example shows the minimum setup to create a {@link DynamoDbSink} that writes
+ * records into DynamoDb
+ *
+ * <pre>{@code
+ * private static class DummyDynamoDbRequestConverter implements DynamoDbRequestConverter<String> {
+ *
+ *         @Override
+ *         public DynamoDbRequest apply(String s) {
+ *             final Map<String, DynamoDbAttributeValue> item = new HashMap<>();
+ *             item.put("your-key", DynamoDbAttributeValue.builder().s(s).build());
+ *             return DynamoDbRequest.builder()
+ *                     .tableName("your-table-name")
+ *                     .putRequest(DynamoDbPutRequest.builder().item(item).build())
+ *                     .build();
+ *         }
+ *     }
+ * DynamoDbSink<String> dynamoDbSink = DynamoDbSink.<String>builder()
+ *                                          .setDynamoDbRequestConverter(new DummyDynamoDbRequestConverter())
+ *                                       .build();
+ * }</pre>
+ *
+ * <p>If the following parameters are not set in this builder, the following defaults will be used:
+ *
+ * <ul>
+ *   <li>{@code maxBatchSize} will be 25
+ *   <li>{@code maxInFlightRequests} will be 50
+ *   <li>{@code maxBufferedRequests} will be 10000
+ *   <li>{@code maxBatchSizeInBytes} will be 16 MB i.e. {@code 16 * 1000 * 1000}
+ *   <li>{@code maxTimeInBufferMS} will be 5000ms
+ *   <li>{@code maxRecordSizeInBytes} will be 400 KB i.e. {@code 400 * 1000 * 1000}
+ *   <li>{@code failOnError} will be false
+ *   <li>{@code dynamoDbTablesConfig} will be empty meaning no records deduplication will be
+ *       performed by the sink
+ * </ul>
+ *
+ * @param <InputT> type of elements that should be persisted in the destination
+ */
+@PublicEvolving
+public class DynamoDbSinkBuilder<InputT>
+        extends AsyncSinkBaseBuilder<InputT, DynamoDbWriteRequest, DynamoDbSinkBuilder<InputT>> {
+
+    private static final int DEFAULT_MAX_BATCH_SIZE = 25;
+    private static final int DEFAULT_MAX_IN_FLIGHT_REQUESTS = 50;
```

Comment:
Am thinking how this will work for users if there are multiple inflight requests with records that have the same PK/SK.
- Seems like it will be a race condition where the first request to complete will take precedence
- If there are failures/partial failures in a batch, the "latest" record in the stream might no longer take precedence. That said, this situation is more likely to occur if there are more records with the same PK/SK, which helps us, because that means eventual consistency will be attained quickly.

In the scenario above, it seems setting `MAX_IN_FLIGHT_REQUESTS` to 1 would be the main mitigation. I wonder if we should set this as default, or at least call this out to users?

Reply:
Very good point, I was thinking about it, I think we should definately add a comment about this. 
We need to be specific here, as if we are using parallelism, MAX_IN_FLIGHT_REQUESTS alone will not help and we need to ensure key and ordering from the source as well
But in general, I think we should document that the sink has this "problem" 
Maybe in the future we can allow to use single PutItem requests and allow expressions as you mentioned in another comment

Reply:
Yes you're right. Users will also need to ensure that a `keyBy` is done before the sink such that all records with same SK/PK goes to the same subtask.

Yep, agreed that we can set the default, and create a JIRA for using single PutItem requests with conditional expressions in the future.

Reply:
I am also thinking, that if we want to really support that we probably have to remove that deduplication that we are doing in case we find a duplicate key, and instead "split" the batch, to make sure we do not lose records with the same primary key

Reply:
@YuriGusev I think it makes sense, what do you think? should we split batches instead of deduplication?

Reply:
Hi, I might be misunderstood the question. 

Duplicated keys only matter when they get to the same batch request, then dynamodb service will fail the request. I took a simple approach to have accumulator to deduplicate entries based on PK/SK in the same request. E.g. if we had 25 entries and 1 entry has duplicated PK/SK we would then write 24 entries in the request.

But if two entries with the same PK/SK get to two different in-flight batch request this is not a problem as deduplication then happens on DynamoDB side.

Reply:
> I am also thinking, that if we want to really support that we probably have to remove that deduplication that we are doing in case we find a duplicate key, and instead "split" the batch, to make sure we do not lose records with the same primary key

Hmm, I think deduplication within a batch on the Flink DDB sink side is a useful feature, because it reduces writes to DDB. The current deduplication ensures the newest record takes precedence, which is the behaviour we see on consecutive `BatchWriteItem` requests.

The difference I can see here is that the ChangeDataCapture stream off the DDB table will be different depending on whether deduping happens in Flink DDB Sink. This might be possible if the user wants to write to DDB but also track number of changes in the stream. This sounds like a relatively niche use case though, since this use case is more suited to a Streaming Destination like Kinesis.

> But if two entries with the same PK/SK get to two different in-flight batch request this is not a problem as deduplication then happens on DynamoDB side.
> If the order of the duplicate/update entries matters I think batch sink can not be used anyway, as the order is mixed up by partial write failures/full retries and parallelism in the sink. Then it is up to user to dedup before the sink.

As far as I can tell, the second request in DDB will take precedence and overwrite the first request. I can see Batching being very useful for Batch jobs where deduping already occured, and we just want to dump a large amt of data to DDB.

For streaming use case, this situation is mitigated if user sets the `MAX_IN_FLIGHT_REQUESTS` to 1, because the `AsyncSinkWriter` will put any failed items from requests to the front of the queue, so the "older" items will be retried first.

In the future, I was thinking we can implement an improvement to use `PutItem` API, and the DDB Sink can add an additional key-value with `eventTime` as the value. Whenever an item with older event time is written, we can just drop the record.

Reply:
May be we could provide a setting to turn off deduplication on the sink side. That would be on-pair with boto3 clients for dynamodb (overwrite_by_pkeys=['partition_key', 'sort_key'] config). Wdyt?

> In the future, I was thinking we can implement an improvement to use PutItem API, and the DDB Sink can add an additional key-value with eventTime as the value. Whenever an item with older event time attempts to be written, we can just drop the older record.

Yes I agree that would be a nice feature to add. We actually had it planned for [the first sink version](https://github.com/apache/flink/pull/17407/files#diff-ffcfaf3e245b27ca44831c37b908014be944626c1eac3d5acc63fa310345d937) (not based on AsyncSyncBase). There was interface that allows different types of producers (batching/non-batching) to be able to extend for this use case in the future. We can introduce something similar here and plan for adding this feature

Reply:
> May be we could provide a setting to turn off deduplication on the sink side

This is the situation if `tableConfig` is not provided right? Or do you think a separate config will be required?

Reply:
Yes this is right if there is no configuration it won't deduplicate anything. It's been a while I forgot I added test for this :D